pkg/util/sketch/series.go

```diff
@@ -0,0 +1,86 @@
+package sketch
+
+import (
+	"github.com/axiomhq/hyperloglog"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/loki/pkg/logproto"
+)
+
+const ValueTypeSketchSeries = "sketches"
+
+type Point struct {
+	topk *Topk
+	ts   uint64
+}
+
+// SketchSeries is `promql.Value` and `parser.Value`
+type Series []Point
+
+// Type implements `promql.Value` and `parser.Value`
+func (Series) Type() parser.ValueType { return ValueTypeSketchSeries }
+
+// String implements `promql.Value` and `parser.Value`
+func (Series) String() string {
+	return ""
+}
+
+func (s Series) ToProto() (*logproto.SketchSeries, error) {
+	points := make([]*logproto.SketchSeries_Point, 0, len(s))
+	for _, point := range s {
+		cms := &logproto.CountMinSketch{
+			Depth: uint32(point.topk.sketch.depth),
+			Width: uint32(point.topk.sketch.width),
+		}
+		cms.Counters = make([]uint32, 0, cms.Depth*cms.Width)
+		for row := uint32(0); row < cms.Depth; row++ {
+			cms.Counters = append(cms.Counters, point.topk.sketch.counters[row]...)
+		}
+
+		hllBytes, err := point.topk.hll.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+
+		topk := &logproto.TopK{
+			Cms:         cms,
+			Hyperloglog: hllBytes,
+		}
+
+		points = append(points, &logproto.SketchSeries_Point{Topk: topk, TimestampMs: int64(point.ts)})
+	}
+
+	return &logproto.SketchSeries{Points: points}, nil
+}
+
+func FromProto(proto *logproto.SketchSeries) (Series, error) {
+	points := make(Series, 0, len(proto.Points))
+	for _, point := range proto.Points {
+		cms := &CountMinSketch{
+			depth: int(point.Topk.Cms.Depth),
+			width: int(point.Topk.Cms.Width),
+		}
+		for row := 0; row < cms.depth; row++ {
+			s := row * cms.width
+			e := s + cms.width
+			cms.counters = append(cms.counters, point.Topk.Cms.Counters[s:e])
+		}
+
+		hll := hyperloglog.New()
+		err := hll.UnmarshalBinary(point.Topk.Hyperloglog)
+		if err != nil {
+			return nil, err
+		}
+
+		// TODO(karsten): should we set expected cardinality as well?
```

Comment:
@cstyan how should we reconstruct the heap? Should we leave it empty and only set the labels?